uxarray/grid/neighbors.py

```diff
@@ -781,6 +781,204 @@ def coordinates(self, value):
+        """
+
+        coords = _prepare_xy_for_query(coords, in_radians, distance_metric=None)
+        bcoords = np.zeros(
+            (coords.shape[0], self._source_grid.n_max_face_nodes), dtype=np.double
+        )
+
+        faces = np.zeros((coords.shape[0]), dtype=INT_DTYPE) - 1  # Default to -1
+
+        # Get the list of faces to search for each coordinate
+        candidate_faces = [
+            self._face_hash_table[pid] for pid in self._hash_index(coords)
+        ]
+        # For each coordinate, perform the in/out check on each candidate face
+        for i, (coord, candidates) in enumerate(zip(coords, candidate_faces)):
+            is_inside = False
+            for face_id in candidates:
+                n_nodes = self._source_grid.n_nodes_per_face[face_id].to_numpy()
+                node_ids = self._source_grid.face_node_connectivity[face_id, 0:n_nodes]
+                nodes = np.column_stack(
+                    (
+                        np.deg2rad(self._source_grid.node_lon[node_ids].to_numpy()),
+                        np.deg2rad(self._source_grid.node_lat[node_ids].to_numpy()),
+                    )
+                )
+                bcoord = _barycentric_coordinates(nodes, coord)
+                err = np.abs(np.sum(bcoord * nodes[:, 0], axis=0) - coord[0])
+                +np.abs(np.sum(bcoord * nodes[:, 1], axis=0) - coord[1])
+                is_inside = all(lambda_i >= 0 for lambda_i in bcoord)
+                if is_inside and err < tol:
+                    break
+
+            if is_inside and err < tol:
+                faces[i] = face_id
+                bcoords[i, 0:n_nodes] = bcoord[0:n_nodes]
+
+        return faces, bcoords
+
+
+def _triangle_area(A, B, C):
+    """
+    Compute the area of a triangle given by three points.
+    """
+    return 0.5 * abs(A[0] * (B[1] - C[1]) + B[0] * (C[1] - A[1]) + C[0] * (A[1] - B[1]))
+
+
+def _barycentric_coordinates(nodes, point):
+    """
+    Compute the barycentric coordinates of a point P inside a convex polygon using area-based weights.
+    So that this method generalizes to n-sided polygons, we use the Waschpress points as the generalized
+    barycentric coordinates, which is only valid for convex polygons.
+
+    Parameters
+    ----------
+        nodes : numpy.ndarray
+            Spherical coordinates (lon,lat) of each corner node of a face
+        point : numpy.ndarray
+            Spherical coordinates (lon,lat) of the point
+    Returns
+    -------
+    numpy.ndarray
+        Barycentric coordinates corresponding to each vertex.
+
+    """
+    n = len(nodes)
+    sum_wi = 0
+    w = []
+
+    for i in range(0, n):
+        vim1 = nodes[i - 1]
+        vi = nodes[i]
+        vi1 = nodes[(i + 1) % n]
+        a0 = _triangle_area(vim1, vi, vi1)
+        a1 = _triangle_area(point, vim1, vi)
+        a2 = _triangle_area(point, vi, vi1)
+        sum_wi += a0 / (a1 * a2)
+        w.append(a0 / (a1 * a2))
+
```

Comment:
For some queries, I'm received a divide by zero error. Can we add a small value to the sum before dividing?

We can import the `ERROR_TOLLERANCE` from `uxarray.constants`

Reply:
Can you share a specific example that is causing this problem so that we can get it added to the tests ? I'd like to make sure whatever patch we put it in here is covered in the test cases .

Reply:
The error arose when I decorated the function with Numba. In it's current state, it raises warnings when using the 30km MPAS grid I tested above, but I haven't run into it in other instances. 

We could extend the existing tests and make sure they don't raise any warnings in the parts this function is called.

Reply:
I've put in the following in the barycentric weights calculation in https://github.com/FluidNumerics/uxarray/commit/378c817ff5911abda8f28679c77671c6031a9ff5

```
        sum_wi += a0 / (max(a1 * a2, ERROR_TOLERANCE))
        w.append(a0 / (max(a1 * a2, ERROR_TOLERANCE)))
```

This clips the denominator to be no smaller than `ERROR_TOLERANCE` from `uxarray.constants` . 

I did not put in an additional test. I'm not familiar with which exact mesh file and which query points you used to produce the error. For a triangle area to be zero (either `a1` or `a2` to be zero), I suspect that a point must lie on an edge of a face or exactly on a corner node. I'm cooking up an example for that at the moment.

Reply:
This is perfect. 

Once you merge the example, feel free to merge, as everything else looks good. We're excited to have this part of our release today.

Reply:
https://github.com/UXARRAY/uxarray/pull/1169/commits/b3c9a0761217f8efc34e57bfd77ffd2d16f03b86 adds tests for querying on vertices and edges in a simple scenario. In this, I found it was needed to clip the individual triangle areas, rather than the denominator as a whole, to get "query on a vertex" to compute the correct value. This should resolve division by zero errors you were encountering with MPAS 30km grid. Once you confirm that, please resolve this conversation so we can get this merged in.